src/resnet.py

```diff
@@ -272,29 +270,34 @@ def _forward_backbone(self, x, return_before_head=False):
         x = self.layer4(x)
         x = self.avgpool(x)
         x = torch.flatten(x, 1)
-
         if self.fc is not None:
-            if return_before_head and (self.pred is None):
-                return x, self.fc(x)
             x = self.fc(x)
+        return x
 
+    def _forward_head(self, x):
         if self.pred is not None:
-            if return_before_head:
-                return x, self.pred(x)
             x = self.pred(x)
-
         return x
 
-    def forward(self, imgs, mc_imgs=None, return_before_head=False):
-        f_imgs = self._forward_backbone(imgs, return_before_head)
-
-        f_mc = None
-        if mc_imgs is not None:
-            f_mc = self._forward_backbone(mc_imgs, False)
-        if (f_mc is not None) or self.return_mc:
-            return f_imgs, f_mc
-
-        return f_imgs
+    def forward(self, inputs, return_before_head=False):
+        if not isinstance(inputs, list):
+            inputs = [inputs]
+        idx_crops = torch.cumsum(torch.unique_consecutive(
+            torch.tensor([inp.shape[-1] for inp in inputs]),
+            return_counts=True,
+        )[1], 0)
+        start_idx = 0
+        for end_idx in idx_crops:
+            _h = self._forward_backbone(torch.cat(inputs[start_idx:end_idx]))
+            _z = self._forward_head(_h)
```

Comment:
Would it be better to let the user decide if they want to use a prediction head? Since without that there are marginal improvements as reported in the paper.

Reply:
User can still use a prediction head  with the ‘--use_pred_head’ flag in `paw_train.py`,  self._forward_head will only run the head if the flag was specified.  We clarified that in `paw_train.py` thanks for the feedback!

Reply:
Oh cool. I guess if the `pred_head` is None then also its effect would be nullified.